packages/apollo-server-codegen/README.md

```diff
@@ -0,0 +1,152 @@
+
+The `[...]Representation` type defaults to `unknown` in non-federated contexts (see below for federation), and is passed as the `parent` property in resolver functions. If the user would like more type safety, they can pass an object to the second type parameter of the emitted `Resolvers` definition, specifying the object name, and the resolvers will have access to those properties when resolving values for that object:
+
+```ts
+const resolvers: Resolvers<TContext, { User: {internalID: number} }> {
+  User: {
+    name({ internalID }) => ... // id will be of type `number`
+  }
+}
+```
+
+Alternatively, if the user would like more type freedom, they can pass `any` to the second type parameter of the emitted `Resolvers` definition, and the resolvers will use `any` as the type of their `parent` argument:
+
+```ts
+const resolvers: Resolvers<TContext, any> {
+  User: {
+    name({ foo, bar }) => ... // this is allowed. `foo` and `bar` will be type `any`.
+  }
+}
+```
+
+### Scalars
+
+Scalars are simply converted to type `any`, though if there is sufficient community demand, these could be specified in a manner similar to the `TInternalReps` strategy above.
+
+### Enums
+
+Enums are by default converted to a basic `or` of all their type options. It is possible to set the `__experimentalInternalEnumSupport` flag to remove type errors when using internal enum values, by simply treating enums as `any`. If there is sufficient community demand, these could be better typed in a manner similar to the `TInternalReps` strategy above.
+
+### Federation Directives
+
+#### @key
+
+Objects with `@key` directives will be translated into resolvers with an additional `__resolveReference` field.
+
+Fields specified as `@key`s will be provided in the `parent` parameter of the each field's resolver definition:
+
+```gql
+type Review @key(fields: "timestamp author { id }") @key(fields: "id") {
+  author: User!
+  id: ID!
+  timestamp: Int!
+}
+
+type User @key(fields: "id") {
+  id: ID!
+  name: String!
+}
+```
+
+```ts
+{
+  Review: {
+    // `representation` is type `{timestamp: number, author: {id: number}} | {id: string}`
+    __resolveReference(representation) {...}
+  }
+  User: {
+    // `representation` is type `{id: string}`
+    __resolveReference(representation) {...}
+    // parent is also {id: string}
+    name(parent) {}
+  }
+}
+```
+
+#### @external, @provides
+
+Fields marked `@external` cannot have resolvers defined for them, unless they are also marked `@provided` by some other field:
+
+```gql
+type Review @key(...) {
+  author: User @provides(fields: "name")
+}
+
+extend type User @key(fields: ID) {
+  id: ID @external
+  name: String @external
+}
+```
+
+```ts
+{
+  User: {
+    id() {...}  // Error
```

Comment:
I don't think this is the correct behavior here. Give a query like this:

```graphql
{
  me {
    reviews {
      id
    }
  }
}
```

Where Review.id is an external field owned by the reviews service, the accounts service needs to be able to provide the `id` field so that the reviews service can make the join. This may mean mapping from `Review.internalID` value in the users service using a custom resolver

Reply:
I thought I allowed fields used in `key`s to be provided always, but I might have missed it. 

Would that be okay, or should I just make it so everything can be provided?

Reply:
Using keys and provides to determine what cant be resolved now.